RFCs/0002-pluggable-discovery.md

```diff
@@ -0,0 +1,568 @@
+# Pluggable Discovery protocol (Version 1.0.0)
+
+## Overview
+
+This document describes how the Pluggable Discovery works and how it should integrate with monitors and uploaders.
+
+## Problem
+
+When a sketch is uploaded to an Arduino board the only way for transferring the binary executable to the microcontroller is through the serial port. Currently:
+
+- the subroutines to enumerate the serial ports and to possibly identify a board are “hardcoded” in the IDE/CLI
+- the only way to identify a board is via USB VID/PID identifiers
+- the "serial monitor", to communicate with the board after the upload, is “hardcoded” in the IDE
+
+The current structure does not allow to use different kind of “ports” to communicate with the microcontroller. For example it would be interesting to have the possibility to:
+
+- discover a board (MDNS) and upload via network (OTA)
+
+- discover a board and upload via a communication port that is not the serial port (for example via USB-HID as implemented by Teensy or via any other kind of port)
+
+- allow a third party to implement his own upload protocol / port
+
+The pluggable discovery aims to provide a solution to these problems.
+
+## Constraints
+
+- A communication port may use any kind of **protocol**
+
+- Software support for a new **protocol** must be added to the system as a **plugin**
+
+- Each port must be **enumerated/discovered**
+
+- Each port may provide **properties relative to the port** (USB serial number / MAC address)
+
+- Each port must have an **unique address and protocol pair**
+
+- A single device may expose multiple ports
+
+## Recommended Solution
+
+The proposed solution is to provide a tool that can enumerate and discover the ports for a **single specific protocol**. There will be a tool for each supported protocol so, in the end, we will have a “Serial ports discovery”, a “Network port discovery” and so on.
```

Comment:
> Even though we specified the port address explicitly, the upload recipe might need additional properties as returned by the discovery tool, so this command must run discovery tools to get that info. Then consider that he AVR platform would both define a serial and network discovery tool, but arduino-cli does not know that the network tool cannot return serial ports, so both must be ran. If e.g. the network LIST command takes a few seconds to wait for mdns replies, this means all serial uploads are delayed by the network discovery, which does not seem acceptable to me.

This is indeed annoying, but the only way to reduce the delay is, as you already suggested, to start the upload as soon as we receive the requested port details from a discovery.

Anyway, this issue is present only in the CLI because the command line flags don't allow to pass a full port object (with all the details) for obvious practicality reasons, so the only ways to overcome this issue are:
1. to query the discoveries again (and possibly wait...)

or

2. to keep a cache of the discovered port

for the IDE this is not a problem because the port selection can be implemented using the cache strategy: the IDE GUI will display to the user only the port name, but internally the IDE will keep the full port details to be used for upload, so we don't need to re-discover the port.

We may think to do something similar in the `arduino-cli`: we cache the full list of ports in a temporary file when we do a `board list` and use the cache file to lookup for a port if the discovery did not reply immediately.

> Alternatively, I realize now, we could add a QUERY command to the protocol, passing a protocol and address, and letting the discovery tool return info in just that one port. This might even be the best solution, since (when implemented properly) could even support custom serial port names (e.g. consider a /dev/serial/my-arduino symlink created by udev, which would not be reported by LIST, but could be resolved by QUERY making it usable) or network hosts that are not discoverable, but can be probed for information.

Adding a `QUERY` command may reduce the traffic between the IDE and the discovery but will not improve the delay, you will always need to wait for the port details to be discovered again.
Also adding it for querying non-listed ports doesn't look correct to me... why the discovery should handle the port in the form `/dev/serial/my-arduino` if it does not list it in the first place? I would like better to change the discovery to handle it directly...

Reply:
I've added some notes about the fact the LIST command may be incomplete and that should be the client (IDE/CLI) to handle it. I think that a QUERY command is not worth it at this moment, but I reserve to add it in revision 2 of the protocol if needed.

Reply:
> Adding a QUERY command may reduce the traffic between the IDE and the discovery but will not improve the delay, you will always need to wait for the port details to be discovered again.

Maybe it can improve the delay, if e.g. you query for a serial port, then the mdns discovery can directly see that it can never discovery that port, so it can return immediately. OTOH, when the cli must do a LIST command, then e.g. the mdns discovery might need some time before it replies, and maybe the cli must wait for all LIST replies to be received.

> Also adding it for querying non-listed ports doesn't look correct to me... why the discovery should handle the port in the form /dev/serial/my-arduino if it does not list it in the first place? I would like better to change the discovery to handle it directly...

Because this `/dev/serial/my-arduino` is just an arbitrary symlink to the actual port and there is no way to autodetect all such symlinks. *if* such symlinks are created through udev, you might be able to query udev for any symlinks it made, but a QUERY command could also support any other symlink manually created by the user even. But you are right that returning such symlinks in the discovery process would be even nicer, and maybe limiting to just udev-generated symlinks is already more than enough.


> I think that a QUERY command is not worth it at this moment, but I reserve to add it in revision 2 of the protocol if needed.

Agreed, an initial implementation will also provide some more insights in whether we would need QUERY and how it should work, so good plan to move forward without it.